File: hw/chavez/u3/hw27Strategy/JavaScript/controller/SortingContext.js
const BubbleSort = require("../model/BubbleSort");
const InsertionSort = require("../model/InsertionSort");
const QuickSort = require("../model/QuickSort");

class SortingContext {

    sort(data) {

        let strategy;

        if (data.length < 30) {
            strategy = new BubbleSort();
        } else if (data.length < 100) {
            strategy = new InsertionSort();
        } else {
            strategy = new QuickSort();
        }

        return strategy.sort(data);
    }

}

module.exports = SortingContext;

File: exams/artieda/u3/Strategy/src/main/java/ec/edu/espe/strategy/model/SortingContext.java
package ec.edu.espe.strategy.model;

public class SortingContext {
    private SortingStrategy sortingStrategy;

    public SortingStrategy setSortStrategy(SortingStrategy sortingStrategy) {
        this.sortingStrategy = sortingStrategy;
        return this.sortingStrategy;
    }

    public SortingStrategy getSortStrategy() {
        return this.sortingStrategy;
    }

    public SortingStrategy setSortStrategyBasedOnSize(int size) {
        if (size < 2) {
            throw new IllegalArgumentException("The number of elements must be greater than 1 (minimum size: 2).");
        }
        if (size >= 2 && size <= 6) {
            this.sortingStrategy = new BubbleSort();
        } else if (size >= 7 && size <= 10) {
            this.sortingStrategy = new InsertionSort();
        } else {
            this.sortingStrategy = new QuickSort();
        }
        return this.sortingStrategy;
    }

    public int[] sort(int[] data) {
        if (data == null) {
            throw new IllegalArgumentException("The input data array cannot be null.");
        }
        if (sortingStrategy == null) {
            setSortStrategyBasedOnSize(data.length);
        }
        return sortingStrategy.sort(data);
    }
}
